File: WebClient SDK/src/default/ui/breadthFirstControlVisitor.ts
import {
  ICardModel,
  IControlViewModel,
  IFormWithBlocksViewModel,
  IBlockViewModel
} from 'tessa/ui/cards';
import {
  TabControlViewModel,
  ContainerViewModel,
  ControlViewModelBase
} from 'tessa/ui/cards/controls';
import { BlockViewModelBase } from 'tessa/ui/cards/blocks';

// tslint:disable:no-any

export abstract class BreadthFirstControlVisitor {
  protected abstract visitControl(control: IControlViewModel);

  protected abstract visitBlock(block: IBlockViewModel);

  public visitByCard(cardModel: ICardModel) {
    const queue: any[] = [];
    for (let block of cardModel.blocks) {
      this.enqueueBlock(block[1], queue);
    }
    for (let form of cardModel.forms) {
      this.enqueueForm(form, queue);
    }

    this.visitInternal(queue);
  }

  public visitByRootControl(rootControl: IControlViewModel) {
    const queue: any[] = [];
    if (rootControl instanceof TabControlViewModel) {
      this.enqueueTabs(rootControl, queue);
    } else if (rootControl instanceof ContainerViewModel) {
      this.enqueueForm(rootControl.form, queue);
    }

    this.visitInternal(queue);
  }

  public visitByForm(form: IFormWithBlocksViewModel) {
    const queue: any[] = [];
    this.enqueueForm(form, queue);
    this.visitInternal(queue);
  }

  public visitByBlock(block: IBlockViewModel) {
    const queue: any[] = [];
    this.enqueueBlock(block, queue);
    this.visitInternal(queue);
  }

  private visitInternal(queue: any[]) {
    while (queue.length !== 0) {
      const item = queue.shift();
      if (item instanceof ControlViewModelBase) {
        this.visitControl(item);

        if (item instanceof TabControlViewModel) {
          this.enqueueTabs(item, queue);
        } else if (item instanceof ContainerViewModel) {
          this.enqueueForm(item.form, queue);
        }
      } else if (item instanceof BlockViewModelBase) {
        this.visitBlock(item);
        this.enqueueBlock(item, queue);
      }
    }
  }

  private enqueueTabs(tabControl: TabControlViewModel, queue: any[]) {
    for (let tab of tabControl.tabs) {
      this.enqueueForm(tab, queue);
    }
  }

  private enqueueForm(form: IFormWithBlocksViewModel, queue: any[]) {
    for (let block of form.blocks) {
      queue.push(block);
    }
  }

  private enqueueBlock(block: IBlockViewModel, queue: any[]) {
    for (let control of block.controls) {
      queue.push(control);
    }
  }
}

File: Source/Extensions/Tessa.Extensions.Default.Client/UI/BreadthFirstControlVisitor.cs
using System.Collections.Generic;
using System.Runtime.CompilerServices;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Controls;

namespace Tessa.Extensions.Default.Client.UI
{
    public abstract class BreadthFirstControlVisitor
    {
        protected abstract void VisitControl(
            IControlViewModel controlViewModel);

        protected abstract void VisitBlock(
            IBlockViewModel blockViewModel);

        public void Visit(
            ICardModel cardModel)
        {
            var queue = new Queue<object>();
            foreach (var block in cardModel.Blocks)
            {
                EnqueueBlock(block.Value, queue);
            }
            foreach (var form in cardModel.Forms)
            {
                EnqueueForm(form, queue);
            }

            this.VisitInternal(queue);
        }

        public void Visit(
            IControlViewModel rootControl)
        {
            var queue = new Queue<object>();
            if (rootControl is TabControlViewModel tabControl)
            {
                EnqueueTabs(tabControl, queue);
            }
            else if (rootControl is ContainerViewModel containerControl)
            {
                EnqueueForm(containerControl.Form, queue);
            }

            this.VisitInternal(queue);
        }

        public void Visit(
            IFormWithBlocksViewModel rootForm)
        {
            var queue = new Queue<object>();
            EnqueueForm(rootForm, queue);
            this.VisitInternal(queue);
        }

        public void Visit(
            IBlockViewModel rootBlock)
        {
            var queue = new Queue<object>();
            EnqueueBlock(rootBlock, queue);

            this.VisitInternal(queue);
        }

        private void VisitInternal(Queue<object> queue)
        {
            while (queue.Count != 0)
            {
                var item = queue.Dequeue();
                if (item is IControlViewModel controlViewModel)
                {
                    this.VisitControl(controlViewModel);

                    if (controlViewModel is TabControlViewModel tabViewModel)
                    {
                        EnqueueTabs(tabViewModel, queue);
                    }
                    else if (controlViewModel is ContainerViewModel containerControl)
                    {
                        EnqueueForm(containerControl.Form, queue);
                    }
                }
                else if (item is IBlockViewModel blockViewModel)
                {
                    this.VisitBlock(blockViewModel);
                    EnqueueBlock(blockViewModel, queue);
                }
            }
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static void EnqueueTabs(TabControlViewModel tabControl, Queue<object> queue)
        {
            foreach (var tabControlTab in tabControl.Tabs)
            {
                EnqueueForm(tabControlTab, queue);
            }
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static void EnqueueForm(
            IFormWithBlocksViewModel formViewModel,
            Queue<object> queue)
        {
            foreach (var blockViewModel in formViewModel.Blocks)
            {
                queue.Enqueue(blockViewModel);
            }
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static void EnqueueBlock(
            IBlockViewModel blockViewModel,
            Queue<object> queue)
        {
            foreach (var controlViewModel in blockViewModel.Controls)
            {
                queue.Enqueue(controlViewModel);
            }
        }


    }
}